module/zfs/dbuf.c

```diff
@@ -3074,12 +3131,25 @@ dbuf_write_ready(zio_t *zio, arc_buf_t *buf, void *vdb)
 		}
 	} else {
 		blkptr_t *ibp = db->db.db_data;
+		int epbs = dn->dn_phys->dn_indblkshift - SPA_BLKPTRSHIFT;
 		ASSERT3U(db->db.db_size, ==, 1<<dn->dn_phys->dn_indblkshift);
 		for (i = db->db.db_size >> SPA_BLKPTRSHIFT; i > 0; i--, ibp++) {
 			if (BP_IS_HOLE(ibp))
 				continue;
 			fill += BP_GET_FILL(ibp);
 		}
+		/*
+		 * it turns out that pre-filling Ln holes in dbuf_read_impl()
+		 * can result in metadata block ids pointing beyond the
+		 * dn_maxblkid; this seems like a good place to fix dn_maxblkid
+		 * up in a manner similar to one above for level-0 blocks
+		 */
+		mutex_enter(&dn->dn_mtx);
+		if (db->db_blkid >
+			dn->dn_phys->dn_maxblkid >> (db->db_level * epbs))
+			dn->dn_phys->dn_maxblkid = ((db->db_blkid + 1) <<
```

Comment:
Hi, @pcd1193182 
thanks for the review. I agree with all the points but would like to discuss the last one about the maxblkid. The reason I added this tweak was to avoid asserts in dbuf_write_done() (see <<< mark):
```
        } else {
                dnode_t *dn;

                DB_DNODE_ENTER(db);
                dn = DB_DNODE(db);
                ASSERT(list_head(&dr->dt.di.dr_children) == NULL);
                ASSERT3U(db->db.db_size, ==, 1 << dn->dn_phys->dn_indblkshift);
                if (!BP_IS_HOLE(db->db_blkptr)) {
                        ASSERTV(int epbs = dn->dn_phys->dn_indblkshift -
                            SPA_BLKPTRSHIFT);
			ASSERT3U(db->db_blkid, <=,
                            dn->dn_phys->dn_maxblkid >> (db->db_level * epbs));  <<<
			ASSERT3U(BP_GET_LSIZE(db->db_blkptr), ==,
                            db->db.db_size);
			if (!arc_released(db->db_buf))
                                arc_set_callback(db->db_buf, dbuf_do_evict, db);
                }
                DB_DNODE_EXIT(db);
                mutex_destroy(&dr->dt.di.dr_mtx);
                list_destroy(&dr->dt.di.dr_children);
        }
```
Your point about the meaning of the maxblkid makes sense. Do you believe it is OK to have Ln metadata blocks that span logical ranges beyond the maxblkid then ? And if so, should this assert be revised ? 

I understand that what the assert says is as follows: for any metadata block, its block id should be such that the upper bound of the logical data (L0) block id range that it spans does not exceed maxblkid. 

Thinking about this for a moment, I believe that even before my changes, the maxblkid could be larger than the highest data block id. For instance, let's take a a file with a single L0 metadata block (pointed to from the dnode). Say all but the top three of the data blocks are filled, and the last three a holes. In order to satisfy the assert above, the maxblkid should be equal to the block id of the top hole. Which is not the last allocated block.